### ISSUE TITLE
train problem

### ISSUE THREAD

**Issue by ClovisChen** [qqwweee/keras-yolo3#85]
I changed train classes to 7 classes, but got a terrible model. 

![test_0 01](https://user-images.githubusercontent.com/23249980/40903612-7d4bec98-680a-11e8-904b-15feecd94484.png)

**Comment by overfitover** [comment]
Do you have  correct labels?

**Comment by bzhong2** [comment]
I have the same problem. The score is very small somehow.

**Comment by qqwweee** [comment]
@ClovisChen 
1. you loaded darknet53 weights? And no freezing layers?
2. you trained 30 epochs? What is train loss and val loss like?

**Comment by ClovisChen** [comment]
@qqwweee 
thank you for your model, sorry I did not see darknet53 weights before. 

1. I converted yolov3.weight to yolo.h5, and loaded it to train. I tried freezing layers and no freezing layers. 
2. After trained 13 epochs, the task is stopped by itself. the train-loss and val loss is 140-150 with the 1024*1024 input shape. Train-loss and val-loss is 30-37 with 512*512 input shape. Train-loss and val-loss is 7-13 with 256*256 input shape. 

 #86 this problem is similar with my trouble. 
I am trying to load darknet53 weights, with freezing layers and no freezing layers. 
I will stick my model performance when I finish my model training.

**Comment by cooli7wa** [comment]
@ClovisChen 

> I converted yolov3.weight to yolo.h5, and loaded it to train.

yolo.h5 is 80 classes, your model is 7 classes, use pretrained yolo.h5 to train your model, is that possible? But loss is so low, why?

**Comment by ClovisChen** [comment]
@cooli7wa 
when yolo.h5 is loaded, the freezing layers is not include the last 3 layers. So I the class numbers can be changed.
my loss is low and stable, but the performance is terrible, the score is 0-0.03, and the bounding boxes are not in the right place. I am trying to figure it out.
![image](https://user-images.githubusercontent.com/23249980/40962722-e9c2ba68-68d8-11e8-98d1-786a049af449.png)

I am trying to load darknet53 weights for training now.

**Comment by ClovisChen** [comment]
@cooli7wa 
`        if freeze_body:
            # => Do not freeze 3 output layers.
            for i in range(len(model_body.layers)-3):
                model_body.layers[i].trainable = False`
when load the pretrained model, the last 3 layers are not loaded.
`def yolo_body(inputs, num_anchors, num_classes):

    """Create YOLO_V3 model CNN body in Keras."""
    darknet = Model(inputs, darknet_body(inputs))
    # => this is the last 3 layers: 1/3
    # => the out_filters is  num_classes + 5
    x, y1 = make_last_layers(darknet.output, 512, num_anchors*(num_classes+5))

    x = compose(
            DarknetConv2D_BN_Leaky(256, (1,1)),
            UpSampling2D(2))(x)
    x = Concatenate()([x,darknet.layers[152].output])
    # => this is the last 3 layers: 2/3
    x, y2 = make_last_layers(x, 256, num_anchors*(num_classes+5))

    x = compose(
            DarknetConv2D_BN_Leaky(128, (1,1)),
            UpSampling2D(2))(x)
    x = Concatenate()([x,darknet.layers[92].output])
    # => this is the last 3 layers: 3/3
    x, y3 = make_last_layers(x, 128, num_anchors*(num_classes+5))

    return Model(inputs, [y1,y2,y3])`
the function `make_last_layers` adapt the num_classes by out_filters. So we can change the num_classes.

**Comment by cooli7wa** [comment]
@ClovisChen 
Thank you,  I need time to think about it.

**Comment by xugaoxiang** [comment]
Need a tutorial about training own dataset(including label tool) seriously for beginers.

**Comment by bzhong2** [comment]
https://github.com/experiencor/keras-yolo3
This repo gave me a much better performance with the same dataset.

**Comment by stefanbo92** [comment]
Hey,
I faced the same issue when training for my own classes: I was training only the last layers (frozen model) and loss went down quickly but the inference results remained quite bad.
It turned out that I just have not trained long enough, so maybe just try out training for ~150 epochs and remove the early stopping. Even if the loss is not decreasing much the model makes some progress. 

I also created a fork wich implements training with bottleneck features. This calculates the output of the frozen layers first (bottleneck features) and then trains only the last layers very quickly (about 30 times faster on my laptop). This makes the training much quicker, even on a CPU, and you can try different training parameters.

Code for it can be found here:
https://github.com/stefanbo92/keras-yolo3

@qqwweee I created a pull request if you like to merge this

**Comment by jyqian-aibee** [comment]
@stefanbo92 thanks for the effort. I tried your code, but it seems it is only available when the dataset size is small. In my case, I have a big dataset, and calculating the last layer features for all of them causes a memory issue and forces my program to be killed. 
Do you have any suggestions on how to get around this?

**Comment by stefanbo92** [comment]
You are right, the bottleneck training works only if you can load all the training data into memory and is especially well suited for smaller datasets. However, you could still precompute the bottlenecks for all your images and instead of saving it as one large file, you can save one bottleneck value set for each image and then rewrite the generator so it will load these "single" bottlenecks.

**Comment by fourth-archive** [comment]
@ClovisChen @overfitover @bzhong2 @qqwweee @cooli7wa @stefanbo92 @jyqian-aibee this YOLOv3 tutorial may help you: 
https://github.com/ultralytics/yolov3/wiki/Train-Custom-Data

The accompanying repository works on MacOS, Windows and Linux, includes multigpu and multithreading, performs inference on images, videos, webcams, and an iOS app. It also tests to slightly higher mAPs than darknet, including on the latest `YOLOv3-SPP.weights` (**60.7 COCO mAP**), and offers the ability to train custom datasets from scratch to darknet performance, all using PyTorch :)
https://github.com/ultralytics/yolov3

<table style="width:100%">
  <tr><th><img src="https://user-images.githubusercontent.com/26833433/54523854-227d0580-4979-11e9-9801-26a3be239875.jpg" width="300"></th>
    <th><img src="https://storage.googleapis.com/ultralytics/logo/logoname1000.png" width="140">
  <br><br/>
  <p> <a href="https://itunes.apple.com/app/id1452689527">
  <img href="https://itunes.apple.com/app/id1452689527" src="https://user-images.githubusercontent.com/26833433/50044365-9b22ac00-0082-11e9-862f-e77aee7aa7b0.png" width="160"> 
  </a> </p></th> 
  </tr>
</table>
